Use copy instead of rename if temp directory is on a different device

bryanwood opened the pull request:
Running `npx "@enhance/create@latest" ./myproject -y` on Windows when the temp directory isn't on the same disk/device as the project directory fails with `Error: EXDEV: cross-device link not permitted`.

Using copySync instead of renameSync fixes the scenario but I'm not sure if the change in this PR is a good idea or not.

bryanwood committed:
use copy instead of rename if temp directory is on a different device

macdonst commented on create-project.js:
```diff
@@ -1,12 +1,15 @@
 import { randomUUID } from 'crypto';
 import { createWriteStream, existsSync, mkdirSync, readFileSync, renameSync, unlinkSync, writeFileSync } from 'fs'
+import fsExtra from 'fs-extra'
```
Should be able to add `copyFileSync` to the `fs` import and delete line 3 where you add the `fs-extra` import.

bryanwood replied:
There is a `cpSync` that does handles copying directories so have used that and removed the fs-extra dependency.

macdonst commented on create-project.js:
```diff
@@ -1,12 +1,15 @@
 import tiny from 'tiny-json-http'
 import tar from 'tar'
 
+const { copySync } = fsExtra;
+
```
You can delete these lines.

macdonst commented on create-project.js:
```diff
@@ -43,8 +46,21 @@ export async function createProject ({ dest, path, name }) {
+
+        if (!isSameFileSystemRoot) {
+            // if not, we need to copy the files instead of moving them
+            copySync(join(temp, 'package'), projectDir)
```
Change to `copyFileSync` and after a successful copy we should remove the temp project folder.

macdonst commented on package.json:
```diff
@@ -22,6 +22,7 @@
     "tape": "^5.6.3"
   },
   "dependencies": {
+    "fs-extra": "^11.1.1",
```
We don't need to add this extra dependency.

macdonst reviewed (changes requested):
Hey @bryanwood thanks for the PR. I'm glad you caught this bug. I'm away from my Windows box right now so I can't fully test this but I have left some comments on the PR that I think will get this in a state in which we can merge.

bryanwood committed:
use built in fs module functions to copy directory and remove temp files

bryanwood committed:
fix formatting

bryanwood commented:
Have addressed the comments and learnt today that node has had APIs for doing recursive directory copying for a while now 😅.

There's another PR #18 that makes the tests run on Windows.

macdonst commented:
@bryanwood this is great, thanks for making those changes. I merged #18 into main. Can you pull those changes into this branch? That way all the tests will run clean on GitHub.

bryanwood committed:
Merge remote-tracking branch 'upstream/main' into use-copy-when-different-filesystem-roots

bryanwood commented:
Looks like `cpSync` is only in node >= 16.7

https://nodejs.org/api/fs.html#fscpsyncsrc-dest-options

macdonst commented:
@bryanwood I've pushed a new branch https://github.com/enhance-dev/create/tree/use-copy-when-different-filesystem-roots which has all your commits plus my own implementation of `cpSync`. Can you pull down the branch and test it on your machine? Seems to be working on mine and in GH Actions but I'd love for the original reporter to validate the fix. Once that's done will commit and cut a new release.

bryanwood commented:
Have tested this locally and it works ^^.

macdonst commented:
@bryanwood thanks for the confirmation. I've pushed the code from my branch with the node 14 copy sync implementation and released an `@enhance/create` v3.1.3 to include the fix.
